File: bench/native_runtime_bench.py
#!/usr/bin/env python3
"""Benchmark real local DSCO modules, including SIGKILL/WAL recovery.

No inference, external messaging, provider substitutes, or saved configuration
changes. All databases and processes belong to this bounded benchmark.
"""
import argparse
import hashlib
import json
import os
from pathlib import Path
import shlex
import signal
import sqlite3
import statistics
import subprocess
import time

ROOT = Path(__file__).resolve().parents[1]


def summary(values):
    ordered = sorted(values)
    return {"count": len(values), "p50": statistics.median(values),
            "p95": ordered[int((len(ordered) - 1) * .95)]}


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--output", type=Path, required=True)
    args = parser.parse_args()
    out = args.output.resolve()
    out.mkdir(parents=True, exist_ok=True)
    if list(out.glob("*.sqlite")):
        parser.error("use a fresh output directory to preserve previous benchmark evidence")
    binary = out / "native_runtime_bench"
    sources = [ROOT / "bench/native_runtime_bench.c"] + [ROOT / f"src/{name}.c"
               for name in ["ipc", "json_util", "env_config", "event_loop", "scheduler"]]
    command = shlex.split(os.environ.get("CC", "cc")) + [
        "-O2", "-std=c11", "-D_DARWIN_C_SOURCE", "-D_POSIX_C_SOURCE=200809L",
        "-I", str(ROOT / "include"), *map(str, sources), "-lsqlite3", "-lm", "-o", str(binary)]
    subprocess.run(command, check=True, timeout=60)
    records = []

    def run(mode, db, count=1000):
        argv = [str(binary), mode, str(db), str(count)]
        start = time.perf_counter()
        result = subprocess.run(argv, capture_output=True, text=True, timeout=30, check=True)
        record = {"mode": mode, "argv": argv, "process_ms": (time.perf_counter() - start) * 1000,
                  "result": json.loads(result.stdout), "stderr": result.stderr}
        records.append(record)
        (out / "records.json").write_text(json.dumps(records, indent=2) + "\n")
        return record

    # Cold = a new native process and IPC connection for every item, reusing DB.
    cold_db = out / "cold.sqlite"
    run("hot", cold_db, 1)  # initialize schema before measuring process restarts
    cold = [run("hot", cold_db, 1) for _ in range(25)]
    hot = [run("hot", out / f"hot-{i}.sqlite", 1000) for i in range(5)]
    queues = run("two-queue", out / "two-queue.sqlite", 1000)
    scheduler = [run("scheduler", "unused") for _ in range(5)]

    recovery_db = out / "recovery.sqlite"
    setup = run("setup", recovery_db)
    holder = subprocess.Popen([str(binary), "hold", str(recovery_db)],
                              stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
    try:
        import selectors
        with selectors.DefaultSelector() as selector:
            selector.register(holder.stdout, selectors.EVENT_READ)
            assert selector.select(timeout=10), "worker failed to claim within deadline"
        line = holder.stdout.readline()
        assert line.strip(), holder.stderr.read()
        claimed = json.loads(line)
        started = time.perf_counter()
        os.kill(holder.pid, signal.SIGSTOP)
        stopped_pid, stopped_status = os.waitpid(holder.pid, os.WUNTRACED)
        assert stopped_pid == holder.pid and os.WIFSTOPPED(stopped_status)
        pause_ms = (time.perf_counter() - started) * 1000
        steering = run("steer", recovery_db)
        os.kill(holder.pid, signal.SIGKILL)
        assert holder.wait(timeout=5) == -signal.SIGKILL
        restarted = run("recover", recovery_db)
        assert restarted["result"]["completed_task"] == setup["result"]["task_id"] == claimed["claimed_task"]
    finally:
        if holder.poll() is None:
            holder.kill()
            holder.wait(timeout=5)
    with sqlite3.connect(recovery_db) as db:
        task = db.execute("SELECT id,status,result,target_agent_id FROM tasks").fetchall()
        checkpoints = db.execute("SELECT COUNT(*) FROM agent_checkpoint").fetchone()[0]
        messages = db.execute("SELECT COUNT(*),COUNT(read_at) FROM messages").fetchone()
        assert len(task) == 1 and task[0][1] == "done" and checkpoints == 1 and messages == (1, 1)

    result = {
        "schema": "dsco.native_runtime_bench.v1",
        "scope": "Native C control-plane operations on local SQLite and cooperative scheduler; no inference or competitor comparison",
        "compiler_argv": command,
        "source_sha256": {str(p.relative_to(ROOT)): hashlib.sha256(p.read_bytes()).hexdigest() for p in sources},
        "cold_process_task_plus_message_ms": summary([r["process_ms"] for r in cold]),
        "cold_ipc_init_ms": summary([r["result"]["init_ms"] for r in cold]),
        "hot_task_lifecycle_p50_ms": summary([r["result"]["task_lifecycle"]["p50_ms"] for r in hot]),
        "hot_task_lifecycles_per_second": summary([r["result"]["task_lifecycle"]["ops_per_second"] for r in hot]),
        "hot_message_roundtrip_p50_ms": summary([r["result"]["message_send_receive_ack"]["p50_ms"] for r in hot]),
        "two_targeted_queues": queues["result"],
        "scheduler_noop_dispatches_per_second": summary([r["result"]["dispatches_per_second"] for r in scheduler]),
        "deferred_callbacks_per_second": summary([r["result"]["callbacks_per_second"] for r in scheduler]),
        "cancelled_task_executions": [r["result"]["cancelled_task_executions"] for r in scheduler],
        "process_recovery": {"sigstop_ms": pause_ms, "claimed": claimed,
            "steering_written_while_stopped": steering["result"], "sigkill_confirmed": True,
            "restarted": restarted, "final_tasks": task, "checkpoints": checkpoints, "messages_total_and_acked": messages},
        "limits": ["No model/inference speed claim", "Cold process timing includes IPC open/schema checks, one task lifecycle, one message roundtrip, cleanup and OS process overhead",
            "Two targeted queues are composed by this harness, not proof of an integrated durable planning scheduler",
            "Task lease is explicitly expired through ipc_task_requeue_stale(0) only after confirmed worker death; no claim of automatic immediate recovery",
            "SQLite WAL uses synchronous=NORMAL: process-crash recovery tested, power-loss durability not tested",
            "Scheduler callbacks are cooperative and process-local; SIGSTOP/SIGKILL exercise OS process interruption",
            "Queued cancellation tested; running callback preemption and completion-attempt fencing are not provided by these measurements",
            "Deferred callbacks are native event-loop callbacks, not HTTP webhook deliveries"],
        "records": records,
    }
    (out / "results.json").write_text(json.dumps(result, indent=2) + "\n")
    print(json.dumps({k: v for k, v in result.items() if k not in ["records", "compiler_argv", "source_sha256"]}, indent=2))


if __name__ == "__main__":
    main()

File: bench/native_runtime_bench.c
/* Native control-plane benchmark: production IPC, scheduler, and event loop.
 * Work items are deliberate no-op payloads; no provider/inference simulation. */
#include "ipc.h"
#include "scheduler.h"
#include "event_loop.h"
#include <assert.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

static double ms(void) {
    struct timespec ts; clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec * 1000.0 + ts.tv_nsec / 1e6;
}
static int cmp(const void *a,const void *b) {
    double x=*(const double *)a,y=*(const double *)b; return (x>y)-(x<y);
}
static void metrics(const char *name,double *samples,int n,double total) {
    qsort(samples,n,sizeof(*samples),cmp);
    printf("\"%s\":{\"count\":%d,\"total_ms\":%.6f,\"ops_per_second\":%.3f,\"p50_ms\":%.6f,\"p95_ms\":%.6f}",
           name,n,total,n*1000.0/total,samples[n/2],samples[(n-1)*95/100]);
}
static int increment(void *ctx) { ++*(int *)ctx; return 0; }
static void callback(void *ctx) { ++*(int *)ctx; }

int main(int argc,char **argv) {
    assert(argc>=3);
    const char *mode=argv[1],*db=argv[2];
    int n=argc>3?atoi(argv[3]):1000; assert(n>0 && n<=100000);
    double start=ms();
    if (!strcmp(mode,"scheduler")) {
        double samples[100]; int completed=0;
        for (int batch=0;batch<100;batch++) {
            scheduler_t scheduler; sched_init(&scheduler); double begin=ms();
            for(int i=0;i<256;i++) assert(sched_spawn(&scheduler,increment,&completed,"native no-op",i%SCHED_PRIO_COUNT)>=0);
            while(sched_tick(&scheduler)>0) {}
            samples[batch]=ms()-begin; sched_destroy(&scheduler);
        }
        double total=ms()-start; assert(completed==25600);
        scheduler_t scheduler; sched_init(&scheduler); int cancelled_calls=0;
        task_id_t id=sched_spawn(&scheduler,increment,&cancelled_calls,"cancelled",SCHED_PRIO_NORMAL);
        int survivor_calls=0;
        assert(sched_spawn(&scheduler,increment,&survivor_calls,"survivor",SCHED_PRIO_LOW)>=0);
        assert(sched_cancel(&scheduler,id)); while(sched_tick(&scheduler)>0) {}
        assert(cancelled_calls==0 && survivor_calls==1);
        assert(sched_task_get(&scheduler,id)->state==TASK_CANCELLED);
        printf("{\"completed\":%d,\"dispatches_per_second\":%.3f,\"cancelled_task_executions\":%d,",completed,completed*1000.0/total,cancelled_calls);
        metrics("batch_256",samples,100,total);
        ev_loop_t *loop=ev_loop_new(); assert(loop); int callbacks=0; start=ms();
        for(int batch=0;batch<100;batch++) {
            for(int i=0;i<64;i++) assert(ev_defer(loop,callback,&callbacks)>=0);
            ev_loop_poll(loop,0);
        }
        assert(callbacks==6400); total=ms()-start;
        printf(",\"deferred_callbacks\":6400,\"callbacks_per_second\":%.3f}\n",6400000.0/total);
        ev_loop_free(loop); return 0;
    }
    char operator_id[64]; snprintf(operator_id,sizeof(operator_id),"operator-%d",getpid());
    const char *identity=(!strcmp(mode,"hold")||!strcmp(mode,"recover"))?"survivor":operator_id;
    assert(ipc_init(db,identity)); double init_ms=ms()-start;
    if (!strcmp(mode,"hot")) {
        assert(ipc_register(NULL,0,"benchmark","local"));
        double *samples=calloc(n,sizeof(double)); assert(samples); start=ms();
        for(int i=0;i<n;i++) {
            double begin=ms(); int id=ipc_task_submit("native no-op",i%5,0); assert(id>0);
            ipc_task_t task; assert(ipc_task_claim(&task)&&task.id==id);
            assert(ipc_task_start(id)); assert(ipc_task_complete(id,"done"));
            samples[i]=ms()-begin;
        }
        printf("{\"init_ms\":%.6f,",init_ms); metrics("task_lifecycle",samples,n,ms()-start);
        start=ms();
        for(int i=0;i<n;i++) {
            double begin=ms(); assert(ipc_send(identity,"steer","continue original objective"));
            ipc_message_t message; assert(ipc_recv(&message,1)==1); free(message.body);
            samples[i]=ms()-begin;
        }
        printf(","); metrics("message_send_receive_ack",samples,n,ms()-start); puts("}"); free(samples);
    } else if(!strcmp(mode,"two-queue")) {
        assert(ipc_agent_define("planner",NULL,0,"planner","","local"));
        assert(ipc_agent_define("executor",NULL,0,"executor","","local")); start=ms();
        for(int i=0;i<n;i++) assert(ipc_task_submit_to("planner","produce execution item",i%5,0)>0);
        ipc_shutdown(); assert(ipc_init(db,"planner"));
        for(int i=0;i<n;i++) {
            ipc_task_t task; assert(ipc_task_claim_targeted(&task));
            assert(ipc_task_start(task.id));
            assert(ipc_task_submit_to("executor","execute planned item",task.priority,task.id)>0);
            assert(ipc_task_complete(task.id,"plan delivered"));
        }
        ipc_shutdown(); assert(ipc_init(db,"executor"));
        for(int i=0;i<n;i++) {
            ipc_task_t task; assert(ipc_task_claim_targeted(&task)); assert(task.parent_task_id>0);
            assert(ipc_task_start(task.id)); assert(ipc_task_complete(task.id,"execution done"));
        }
        double total=ms()-start;
        printf("{\"planning_items\":%d,\"execution_items\":%d,\"total_ms\":%.6f,\"pipelines_per_second\":%.3f}\n",n,n,total,n*1000.0/total);
    } else if(!strcmp(mode,"setup")) {
        assert(ipc_agent_define("survivor",NULL,0,"durable benchmark","","local"));
        int id=ipc_task_submit_to("survivor","survive interruption",10,0); assert(id>0);
        printf("{\"task_id\":%d}\n",id);
    } else if(!strcmp(mode,"hold")) {
        assert(ipc_agent_activate()); ipc_task_t task; assert(ipc_task_claim_targeted(&task));
        assert(ipc_task_start(task.id));
        assert(ipc_checkpoint_save(identity,1,"{\"step\":1}","[]","{\"objective\":\"finish\"}","{\"phase\":\"claimed\"}"));
        printf("{\"pid\":%d,\"claimed_task\":%d,\"checkpoint\":1}\n",getpid(),task.id); fflush(stdout);
        for(;;) pause();
    } else if(!strcmp(mode,"steer")) {
        assert(ipc_send("survivor","interrupt","resume from checkpoint and finish"));
        puts("{\"message_persisted\":true}");
    } else if(!strcmp(mode,"recover")) {
        /* The driver has proved process death before expiring this lease. */
        start=ms(); int requeued=ipc_task_requeue_stale(0); assert(requeued==1);
        char *memory=NULL,*plan=NULL;
        assert(ipc_checkpoint_restore(identity,-1,&memory,NULL,&plan,NULL));
        assert(!strcmp(memory,"{\"step\":1}")&&!strcmp(plan,"{\"objective\":\"finish\"}")); free(memory); free(plan);
        ipc_message_t message; assert(ipc_recv(&message,1)==1); assert(!strcmp(message.topic,"interrupt")); free(message.body);
        ipc_task_t task; assert(ipc_task_claim_targeted(&task)); assert(!strcmp(task.target_agent_id,identity));
        assert(ipc_task_start(task.id)); assert(ipc_task_complete(task.id,"resumed and completed"));
        printf("{\"requeued\":%d,\"checkpoint_restored\":true,\"message_survived\":true,\"completed_task\":%d,\"recovery_ms\":%.6f,\"init_ms\":%.6f}\n",requeued,task.id,ms()-start,init_ms);
    } else { fprintf(stderr,"unknown mode\n"); return 2; }
    ipc_shutdown(); return 0;
}
